lib/mqtt.js

```diff
@@ -66,7 +66,7 @@ class MQTTClient {
                 agent: getWSProxyAgent(this.config.brokerURL)
             }
         }
-        this.client = mqtt.connect(brokerURL, this.brokerConfig)
+        this.client = mqtt.connect(brokerURL.href, this.brokerConfig)
```

Comment:
As far as I can see, this will cause the ERR_INVALID_URL error mentioned in the code above.

TLDR;
Using `brokerURL.href` results in sending the original URL (e.g. 'ws://[::1]:8883/'`) which gets manually deconstructed and reconstructed in the bowels of the MQTT lib as 'ws://::1:8883/'` resulting in the original problem.


DETAILS:

REF: https://github.com/mqttjs/MQTT.js/issues/1569

I have looked at the src for V5 and they [still use](https://github.com/mqttjs/MQTT.js/blob/accd78e38aa82c8cc1ea04029e56494276776c87/lib/connect/index.js#L64) the depreciated `url.parse` to deconstruct the URL then reconstruct it manually which is what out workaround was fixing.


The only thing I am not 100% certain of is if the `import url` at the top of the v5 code causes the code to use the old depreciated `url.parse` or the new `new URL` methods after the ts is compiled.  If it is the latter, the problem goes away.



To confirm this I will need to pull this branch and do local testing with IPV6

Reply:
Unfortunately, this does indeed cause an error when a IPV6 IP URL is used.

Looking through the `MQTT.js` src, i cannot see how passing a string vs passing the URL object makes any difference to reconnect options.

If I have missed something there may be an alternative route. If we must send a string URL then there is an opportunity to intercept and update the final connection URL by setting the connect opts function `transformWsUrl` 

NOTE: I have raised a PR against the linked issue on `MQTT.js`

Reply:
Passing the object was leading to it trying to  connect to localhost in the device agent even when the URL pointed to staging.